### ISSUE TITLE
Meta: Suggestions for issue labels

### ISSUE THREAD

**Issue by JedWatson** [keystonejs/keystone#1928]
Throwing this out there for feedback.

At the moment, there are a bunch of issue labels in our project and I don't think we're using them very effectively. I'd like to see a smaller set that are actually useful and apply them more consistently.

Does anybody have any feedback on what should be included / excluded? Please let me know by suggesting:

- a list of suggested labels and uses (start from scratch if you like)
- current labels that you think are *useful*
- current labels you think are *not useful*
- labels you think should be added

Thanks!

**Comment by JedWatson** [comment]
I should mention that when the labels get sorted out we'll add explanations of what they mean / intended use to the wiki so new contributors can understand them. At the moment, some are really obscure (e.g. the "roadmap" label) and could mean several different things.

**Comment by ericelliott** [comment]
Here is the set we're currently using for JSHomes:

<img width="181" alt="screen shot 2015-07-09 at 12 37 44 pm" src="https://cloud.githubusercontent.com/assets/364727/8603760/7fdf8374-2637-11e5-9453-f6d9aefb8d04.png">

Some of these are designed to help people find things that they have the expertise to contribute to, e.g.:

* WordPress (obviously not applicable to this project)
* Universal-JS
* Design

I'm not using "Won't Fix" -- I find it's a bit redundant with simply closing the issue.

I've also recently been using green for "help wanted". The idea being we want to give people the "green light" to pitch in. ;)

**Comment by ericelliott** [comment]
Here's a simplified set that only includes labels I frequently use (with colors I like for them):

<img width="219" alt="screen shot 2015-12-06 at 6 02 19 pm" src="https://cloud.githubusercontent.com/assets/364727/11616886/a5ab56c8-9c43-11e5-9e4d-fe9d9051835e.png">

**Comment by morenoh149** [comment]
`help wanted` sounds equivalent to `easy for beginner` which is the only one I'd like to mention. I think it's great to point out easy tasks potential new contributors can pick up. But if no one else feels the same way we could remove it.

the translation tags were useful once upon a time but I'm not sure how committed we all are to syncing translations.

`enhancement` and `new feature` should definitely be consolidated.
`duplicate` should be removed as duplicate issues are usually closed

**Comment by creynders** [comment]
Ideally the issue labels communicate the following things:

- The state of the issue, i.e. progress towards a solution
- The type of issue (question, bug...)
- Which tier, system or domain part the issue relates to
- The severity of the issue
- Any other important or useful information

I'll clarify and give my proposals: 

**0/ severity**: determining the severity is real hard, especially when there's no real roadmap, so I think we should assume by default the severity is "normal" and only add one: `0-critical`. There's lots of terms used, e.g. "blocker", but I don't like it, it sounds too much waterfall-methodology to me, and the aim should be to reach (nearly) continuous releases, in which case "blocker" makes no sense. This label should be used in one case only, it needs to be fixed ASAP (e.g. the issue about being able to login with "." as a user name)

- "0-critical"

**1/ issue type**: there's many, I'd go for:

- "1-question"
- "1-bug"
- "1-request" (request for a feature/enhancement)
- "1-feature" (PR adding a new feature) Or we could drop "1-request", i.e. if it's an issue it's a feature request, if it's a PR it's a feature addition
- "1-discussion" (API, architectural or security measure discussions)
- "1-cleanup" (refactorings, removal of obsolete files etc)

**2/ issue state**:

- "2-ready" to be merged
- "2-needs-input" more info was asked, or some problematic commits were identified, a rebase agains master is needed i.e. the creator needs to _do_ something
- "2-postponed" the PR can be considered once another PR has landed, or something else which withholds further progress. Applicable to feature requests too
- "2-duplicate" closed due to duplicate
- "2-wont-fix" closed, since it was considered a bad idea or it's WAI
- "2-confirmed" a bug is confirmed
- "2-in-progress" WIP
- "2-needs-review"
- "2-needs-confirmation" (Maybe "2-needs-review" is enough?)

Any open issues w/o a state label are in need of triage and any closed issues w/o a state label are "fixed/merged/solved/landed".

**3/ system/domain/tier part**: there's many directions we could go with this one, but I think these make most sense

- "3-UX"
- "3-security"
- "3-meta" E.g. we need to adhere to json-schema! or We should switch to koa!
- "3-docs"
- "3-build" browserify, babelify, gulp, etc. etc.
- "3-QA" eslint, tests, ...

**4/ Anything useful**:

- "4-needs-help"
- "4-easy" for beginners
- "4-react" this would be useful, since it's an important part that you need very specific knowledge for to contribute

IMO an non-fixed issue should always close with a label, since it allows for an easy overview of why an issue was closed, which is why I _do_ think we should keep `duplicate`. 

The logic behind the groups is that the higher the number, the fuzzier it gets and the potential to needing several labels from the same group becomes bigger, i.e. 0, 1 and 2 are almost certainly exclusive. 3 probably, but not necessarily and 4 not at all.

**Comment by ericelliott** [comment]
> There's lots of terms used, e.g. "blocker", but I don't like it, it sounds too much waterfall-methodology to me, and the aim should be to reach (nearly) continuous releases, in which case "blocker" makes no sense.

In my issue tags, "blocked" means that the issue is waiting for the resolution of another issue before it can be resolved. "blocking" means that an issue is blocking the resolution of another issue. Blocking issues naturally get a bit more priority, because they're holding up the progress of other issues.

This has nothing to do with waterfall. It has to do with identifying anything that's holding up progress so we can clear them as quickly as possible.

Priority tags tend to get fuzzy.  Different issues have different priorities for different people. I prefer to group tickets into milestones. Are we working on it this milestone, or not -- that's all I really care about.

Your other suggestions sound very organized. I've tried many systems like this. IMO, lots of tags mostly just add management overhead without making the process faster/better.

**Comment by stkrzysiak** [comment]
I agree with @morenoh149 about the importance of having a tag for beginner friendly issues.  I don't see one for 'review needed', so I'll throw that out there. 
@ericelliott What is the smoke-test label for?

**Comment by ericelliott** [comment]
> @ericelliott What is the smoke-test label for?

It means that a change needs some manual smoke tests before release.

**Comment by webteckie** [comment]
From a process flow perspective I tend to see it as:

1. type: (bug | question | enhancement)
2. urgency: (blocking)
3. expertise: (easy-for-beginner | advanced-knowledge)
4. state: (help-wanted | not-assigned | assigned) -> in-progress -> (blocked | resolved) -> (needs-review | needs-test | ready-for-merge | works-as-expected | wont-fix | duplicate) -> closed

For example, I could see the following groupings of labels applied at any one time:
bug, blocking, advanced-knowledge, help-wanted
bug, blocking, advanced-knowledge, not-assigned
bug, advanced-knowledge, assigned, in-progress
bug, easy-for-beginner, resolved, needs-review
bug, advanced-knowledge, resolved, ready-for-merge
bug, advanced-knowledge, works-as-expected, closed

Any extra info can be conveyed via commenting.  In any case, I suppose the more complicated issue would be managing the labels :-)

**Comment by ericelliott** [comment]
> not-assigned | assigned

Handled by actual assignments. No need to recreate these features in tags.

> in-progress

That one's useful. =)

> blocked | resolved

The `resolved` state is implied if it's not `blocked`, right?

> needs-review

Implied by an open pull request

> needs-test | ready-for-merge

Useful.

> works-as-expected | wont-fix | duplicate

I find all of these redundant with closing the ticket. I find that trying to track these things adds more project management overhead than it solves.

=)

**Comment by webteckie** [comment]
Re: the not-assigned/assigned...yeah, right ... scrap that...

Re: blocked vs resolved...but an issue may become blocked before it is resolved during dev for whatever reason (in comments)...then at some point in the future it becomes unblocked and continues in-progress and eventually gets resolved.

Re: needs-review...I guess :-)

Re: works-as-expected | ...yeah, I guess so as well :-)

**Comment by ericelliott** [comment]
> then at some point in the future it becomes unblocked and continues in-progress and eventually gets resolved.

And then you remove the `blocked` label, right? Not sure I see a need for `resolved`. In fact, I think that would just cause some weird questions, like "what was resolved, exactly?" & "Why do we care about it anymore if it's resolved?"

**Comment by creynders** [comment]
@ericelliott re.: blocked/blocking that's entirely different from "blocker", which is a JIRA thing if I remember correctly. Sounds good to me! Just wondering whether this won't lead to a bunch of "blocked" issues that never got updated after a "blocker" was resolved.

>Priority tags tend to get fuzzy.

I fully agree. Which is why I only suggest one, for _really_ REALLY urgent ones.

And I agree on "resolved", I don't see a need for it.

I definitely think we have to limit the number of tags to as few as possible.

**Comment by ericelliott** [comment]
> Just wondering whether this won't lead to a bunch of "blocked" issues that never got updated after a "blocking" was resolved.

Yeah, there will be process that goes with these tags. You need to link to the blocking/blocked issue. When the status changes, update both issues.

I believe the "blocker" tag you're thinking of from Jira may be "release blocker". I don't think that tag is needed. If it's in the milestone, it's a release blocker. If it's not a release blocker anymore, and it's time for a release, move the uncompleted issue to a different milestone, instead of tagging it.

> I fully agree. Which is why I only suggest one, for really REALLY urgent ones.

I'm on board with that, as long as there is only one tag: "urgent" sounds good. =)

**Comment by LorbusChris** [comment]
Additionally to any of the proposed suggestions I think it might be useful to give the issues some context as to what part of keystone they concern. Thinking of this slide in Jed's Nodevember speech:
https://www.youtube.com/watch?v=of2VM3Wwkxk&t=27m15s

* List Schema
* Admin UI
* App
* REST
* GraphQL
* Mongoose
* Any other DB

@JedWatson maybe put the slide up on the roadmap to give everyone a good understanding of what the targeted architecture looks like?

**Comment by morenoh149** [comment]
React native uses "Good first task" for the same idea I brought up before.

**Comment by mxstbr** [comment]
Based on reading through the discussion above, I think this is the currently preferred label list:

`critical`
`question`, `bug`, `enhancement`, `discussion`, `good first task`
`in progress`, `needs review`, `needs input`, `blocked`, `on hold`

If we opt for domain-specific labels (which I'm not sure are useful), I think the current system is alright but a bit too elaborate. I'd shorten it to:

`admin`, `fields`, `core`, `docs`

The `security` label gets a fair amount of use at the moment but might be replaced by the `critical` label.

Does that look good to everybody?

**Comment by mxstbr** [comment]
Implemented the above labels.
